### PR TITLE
chore: update `spec-new` in CONTRIBUTING.md [skip ci]

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| License                  | MIT

This PR updates the process of developing a new babel plugin in this repo. It features a detailed checklist and an updated example of a new parser feature PR.

The process also requires an AST PR addressed in ESTree before implementation. By doing so we can gather more feedback, especially about the AST shape, before a new proposal is implemented in Babel. In the future we will follow ESTree unless the new AST is _not_ consistent with current Babel AST.

/cc @nzakas @RReverser 

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/11706"></a>

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
> In the future we will follow ESTree unless the new AST is not consistent with current Babel AST.

Also not if the experimental estree nodes change after that we implemented them (and they are free to do so).

**Comment by JLHwung** [comment]
@nicolo-ribaudo Though it is called experimental in ESTree, the AST changes is still rare once it gets approved: https://github.com/estree/estree/commits/master/experimental

We should document our AST stability policy somewhere. Here it works as a rough guideline to new contributors working on new language features.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -311,15 +311,22 @@ Note that the code shown in Chrome DevTools is compiled code and therefore diffe
...
+> Example: https://github.com/babel/babel/pull/11640
 
-- Create a new issue that describes the proposal (ex: [#538](https://github.com/babel/babylon/issues/538)). Include any relevant information like proposal repo/author, examples, parsing approaches, meeting notes, presentation slides, and more.
-- The pull request should include:
-  - [ ] An update to the [plugins](https://github.com/babel/babel/tree/main/packages/babel-parser#plugins) part of the readme. Add a new entry to that list for the new plugin flag (and link to the proposal)
-  - [ ] If any new nodes or modifications need to be added to the AST, update [ast/spec.md](https://github.com/babel/babel/blob/main/packages/babel-parser/ast/spec.md)
-  - [ ] Make sure you use the `this.hasPlugin("plugin-name-here")` check in the babel parser so that your new plugin code only runs when that flag is turned on (not default behavior)
+- Create a new PR that describes the proposed AST shape in [ESTree](https://github.com/estree/estree) (ex: [Decimal AST](https://github.com/estree/estree/pull/220)). The new AST should follows ESTree's [design philosophy](https://github.com/estree/estree#philosophy).
```
Comment (existentialism) [inline CONTRIBUTING.md:316]:
```suggestion
- Create a new PR that describes the proposed AST shape in [ESTree](https://github.com/estree/estree) (ex: [Decimal AST](https://github.com/estree/estree/pull/220)). The new AST should follow ESTree's [design philosophy](https://github.com/estree/estree#philosophy).
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -311,15 +311,22 @@ Note that the code shown in Chrome DevTools is compiled code and therefore diffe
...
-  - [ ] If any new nodes or modifications need to be added to the AST, update [ast/spec.md](https://github.com/babel/babel/blob/main/packages/babel-parser/ast/spec.md)
-  - [ ] Make sure you use the `this.hasPlugin("plugin-name-here")` check in the babel parser so that your new plugin code only runs when that flag is turned on (not default behavior)
+- Create a new PR that describes the proposed AST shape in [ESTree](https://github.com/estree/estree) (ex: [Decimal AST](https://github.com/estree/estree/pull/220)). The new AST should follow ESTree's [design philosophy](https://github.com/estree/estree#philosophy).
+- After the ESTree PR is accepted, update [ast/spec.md](https://github.com/babel/babel/blob/master/packages/babel-parser/ast/spec.md). Note that there are differences between Babel AST and ESTree. In these cases, consistency with current Babel AST outweighs alignment to ESTree. Otherwise it should follow ESTree.
+
+- [ ] Implement parser plugins based on the new AST. Name parser plugin by camelizing the unprefixed slug of the TC39 proposal URL. i.e. `exportDefaultFrom` from `https://github.com/tc39/proposal-export-default-from`.
```
Comment (existentialism) [inline CONTRIBUTING.md:319]:
```suggestion
- [ ] Implement parser plugins based on the new AST. The name parser plugin should be a the unprefixed slug of the TC39 proposal URL in _camelcase_, i.e. `exportDefaultFrom` from `https://github.com/tc39/proposal-export-default-from`.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -311,15 +311,22 @@ Note that the code shown in Chrome DevTools is compiled code and therefore diffe
...
   - [ ] Add failing/passing tests according to spec behavior
-- Start working about the Babel transform itself!
+  - [ ] Add `@babel/syntax-new-syntax` package. You can copy `packages/babel-plugin-syntax-decimal` and replace `decimal` to `new-syntax`.
+  - [ ] Add `@babel/syntax-new-syntax` to `@babel/standalone`.
+    - [ ] Add `@babel/syntax-new-syntax` to `package.json`
+    - [ ] Add `@babel/syntax-new-syntax` to [`pluginsConfig.json`](https://github.com/babel/babel/blob/master/packages/babel-standalone/scripts/pluginConfig.json), run `make generate-standalone`.
+    - [ ] Add `@babel/syntax-new-syntax` to `src/preset-stage-x`.
+  - [ ] Add `"newSyntax"` to parser [typings](https://github.com/babel/babel/blob/master/packages/babel-parser/typings/babel-parser.d.ts)
+- [ ] Implement generator support in `packages/babel-generator/src/generators`. The generator converts AST to source code.
+- [ ] If this feature can be transpiled, start working about the Babel transform.
```
Comment (existentialism) [inline CONTRIBUTING.md:329]:
```suggestion
- [ ] If this feature can be transpiled, start working on the Babel transform.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -311,15 +311,22 @@ Note that the code shown in Chrome DevTools is compiled code and therefore diffe
...
+- Create a new PR that describes the proposed AST shape in [ESTree](https://github.com/estree/estree) (ex: [Decimal AST](https://github.com/estree/estree/pull/220)). The new AST should follow ESTree's [design philosophy](https://github.com/estree/estree#philosophy).
+- After the ESTree PR is accepted, update [ast/spec.md](https://github.com/babel/babel/blob/master/packages/babel-parser/ast/spec.md). Note that there are differences between Babel AST and ESTree. In these cases, consistency with current Babel AST outweighs alignment to ESTree. Otherwise it should follow ESTree.
+
+- [ ] Implement parser plugins based on the new AST. Name parser plugin by camelizing the unprefixed slug of the TC39 proposal URL. i.e. `exportDefaultFrom` from `https://github.com/tc39/proposal-export-default-from`.
+  - [ ] Use the `this.expectPlugin("newSyntax")` check in `@babel/parser` so that your new plugin code only runs when that flag is turned on (not default behavior), and it will throw friendly errors when users forget to enable a plugin.
```
Comment (existentialism) [inline CONTRIBUTING.md:320]:
```suggestion
  - [ ] Use the `this.expectPlugin("newSyntax")` check within `@babel/parser` to ensure your new plugin code only runs when that flag is turned on (not default behavior), and a friendly error is thrown if users forget to enable a plugin.
```

**Comment by JLHwung** [comment]
> it would be nice to maybe document where/when we do things somewhere (I used to do it at https://github.com/babel/proposals, example like babel/proposals#18) so you can track history but it's been some work to keep it all organized/updated

Yeah a repo / issue tracking each proposals is always good-to-have but proves a bit overhead, at it is prone to become outdated compared to https://github.com/tc39/proposals.

Currently we use labels to track issues related to specific proposals. It is easier to maintain rather than updating stages/status manually in `babel/proposals`.

**Pull request merged**